src/handleCommand.ts

```diff
@@ -1,41 +1,40 @@
-import { Handler, InbuiltCommand } from "./classes";
-import chalk from "chalk";
+import { spawn } from "child_process";
 import { readdirSync } from "fs";
-import { shellFlags } from ".";
-import { spawn, execSync } from "child_process";
+import { Handler, InbuiltCommand } from "./classes";
+import { ExitCodes } from './constants';
+import { findInPath } from "./util";
 export class CommandHandler extends Handler {
   static async prepare() {
     for (let file of readdirSync(__dirname + "/commands")) {
       const data = await import(__dirname + "/commands/" + file);
       file = file.replace(/\.js$/, "");
-      if (shellFlags.flags.debug)
-        process.stdout.write(chalk`Loaded inbuilt command {bold ${file}}\n`);
       this.inbuiltCommands[file] = data.default;
     }
   }
-  static invoke(data: string) {
-    const rawArgs = data.split(/\s+/),
-      args = rawArgs.slice(1),
-      commandName = rawArgs[0];
-    if (CommandHandler.isSyntaxError(data))
-      return { out: "syntax error: unexpected end of input\n", code: 127 };
+  static async invoke(args: string[], variables: Record<string, string>): Promise<{ out: string, code: NodeJS.Signals | ExitCodes }> {
+    let commandName = args.shift();
     if (commandName in CommandHandler.inbuiltCommands)
       return new CommandHandler.inbuiltCommands[commandName]()
-        .prepare(this, args)
+        .prepare(this, args, variables)
         .invoke();
     try {
-      execSync("which " + commandName);
-    } catch {
-      return { out: `${commandName}: command not found\n`, code: 127 };
+      const input = commandName;
+      commandName = await findInPath(commandName);
+      if (!commandName) return { out: `${input}: command not found\n`, code: ExitCodes.COMMAND_NOT_FOUND };
+    } catch (err) {
+      if (typeof err !== "string") return { out: `An unknown error occurred`, code: ExitCodes.UNKNOWN_ERROR }
+      return { out: err, code: ExitCodes.ERROR };
     }
-    return new Promise<{ out: ""; code: NodeJS.Signals | number }>((r) => {
-      const child = spawn(commandName + " " + args.join(" "), { shell: true });
+    return new Promise((r) => {
+      const child = spawn(
+        `${Object.entries(variables)
+          .map(([key, value]) => `${key}="${value}"`)
+          .join(" ")} ${commandName} ${args.map(arg => `"${arg}"`).join(" ")}`,
+        { shell: true }
+      );
       child.stdout.on("data", (d) => process.stdout.write(d.toString()));
       child.on("exit", (code, signal) => r({ out: "", code: code || signal }));
     });
   }
-  static isSyntaxError(str: string) {
-    return /(^&$|^function\s*$|^fn\s*$)/.test(str);
-  }
```

Comment:
I don't really know why I put these there even though I didn't even have a `fn` command,  might add a TODO comment there

Reply:
yeah I was kinda confused by this so I just removed it lmao. what would a fn command be like? we could just allow Javascript to be entered and eval it (in a different context so it doesn't affect splatsh internals)